lib/uri_checker.rb

```diff
@@ -0,0 +1,213 @@
+require "uri"
+
+module UriChecker
+  class Report
+    attr_reader :errors, :warnings
+
+    def initialize(errors: nil, warnings: nil)
+      @errors = errors || Hash.new { |hash, key| hash[key] = [] }
+      @warnings = warnings || Hash.new { |hash, key| hash[key] = [] }
+    end
+
+    def merge!(other)
+      errors.merge!(other.errors) { |key, oldval, newval| oldval | newval }
+      warnings.merge!(other.warnings) { |key, oldval, newval| oldval | newval }
+    end
+
+    def has_errors?
+      !errors.empty?
+    end
+  end
+
+  class ValidUri
+    HTTP_URI_SCHEMES = %w(http https)
+    FILE_URI_SCHEMES = %w(file)
+
+    attr_reader :report, :options
+
+    def initialize(options = {})
+      @report = Report.new
+      @options = options
+    end
+
+    def call(uri)
+      parsed_uri = URI.parse(uri)
+
+      if parsed_uri.scheme.nil?
+        report.warnings[:no_scheme] << "No scheme given."
+      elsif HTTP_URI_SCHEMES.include?(parsed_uri.scheme)
+        report.merge!(HttpChecker.new(parsed_uri, options).call)
+      elsif FILE_URI_SCHEMES.include?(parsed_uri.scheme)
+        report.merge!(FileChecker.new(parsed_uri, options).call)
+      else
+        report.warnings[:unsupported_scheme] << "Unsupported scheme."
+      end
+
+      report
+    rescue URI::InvalidURIError
+      report.errors[:uri_invalid] << "Invalid URI"
+      report
+    end
+  end
+
+  class HttpChecker
+    INVALID_TOP_LEVEL_DOMAINS = %w(xxx adult)
+    REDIRECT_LIMIT = 10
+    REDIRECT_WARNING = 2
+    TIMEOUT = 5
+    SLOW_RESPONSE_TIME = 2
+
+    attr_reader :uri, :redirect_history, :report
+
+     def initialize(uri, redirect_history: [])
+      @uri = uri
+      @redirect_history = redirect_history
+      @report = Report.new
+    end
+
+    def call
+      check_redirects
+      return report if report.has_errors?
```

Comment:
this seems premature as you could have a redirect with the below problems (unless I'm getting something wrong)